### PR TITLE
[Arm64] GetLargestOnDieCacheSize

### PR DESCRIPTION
**Pull request by sdmaclea**
Look for cache info in /sys/.../cache/index*/size

If that fails estimate cache size based on processor count

### CODE REVIEW

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -35488,7 +35488,7 @@ size_t GCHeap::GetValidGen0MaxSize(size_t seg_size)
         // performance data seems to indicate halving the size results
         // in optimal perf.  Ask for adjusted gen0 size.
         gen0size = max(GCToOSInterface::GetLargestOnDieCacheSize(FALSE)/GCToOSInterface::GetLogicalCpuCount(),(256*1024));
-#if (defined(_TARGET_AMD64_))
+
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: Assume arm64 should also limit gen0 to 1/6 of physical memory.

File: src/pal/src/misc/sysinfo.cpp [hunk 2]
```diff
@@ -384,6 +384,53 @@ PAL_HasGetCurrentProcessorNumber()
     return HAVE_SCHED_GETCPU;
 }
 
+bool
+PALAPI
+PAL_ReadMemoryValueFromFile(const char* filename, size_t* val)
```
Comment (sdmaclea) [inline src/pal/src/misc/sysinfo.cpp]: @janvorli This is a cut/paste rename from `CGroup::ReadMemoryValueFromFile() `  Seems wrong to create a third clone of this function.  Whare should it go.

File: src/pal/src/misc/sysinfo.cpp [hunk 3]
```diff
@@ -403,5 +450,47 @@ PAL_GetLogicalProcessorCacheSizeFromOS()
...
+        if(PAL_ReadMemoryValueFromFile("/sys/devices/system/cpu/cpu0/cache/index0/size", &size))
+            cacheSize = max(cacheSize, size);
+        if(PAL_ReadMemoryValueFromFile("/sys/devices/system/cpu/cpu0/cache/index1/size", &size))
+            cacheSize = max(cacheSize, size);
+        if(PAL_ReadMemoryValueFromFile("/sys/devices/system/cpu/cpu0/cache/index2/size", &size))
+            cacheSize = max(cacheSize, size);
+        if(PAL_ReadMemoryValueFromFile("/sys/devices/system/cpu/cpu0/cache/index3/size", &size))
+            cacheSize = max(cacheSize, size);
+        if(PAL_ReadMemoryValueFromFile("/sys/devices/system/cpu/cpu0/cache/index4/size", &size))
+            cacheSize = max(cacheSize, size);
```
Comment (sdmaclea) [inline src/pal/src/misc/sysinfo.cpp]: @janvorli Hardcoded paths, philisophically sysfs might be mounted on a different mount point, but this seems to be the cannonical path.

Some of these index correspond to ICache, but they would be small caches anyway, so I ignored that.

index4 should be L4$ which seems big enough for current processors.

File: src/vm/util.cpp [hunk 4]
```diff
@@ -1861,8 +1861,6 @@ size_t GetLargestOnDieCacheSize(BOOL bTrueSize)
     STATIC_CONTRACT_NOTHROW;
     STATIC_CONTRACT_GC_NOTRIGGER;
 
-#if defined(_TARGET_AMD64_) || defined (_TARGET_X86_)
```
Comment (sdmaclea) [inline src/vm/util.cpp]: Share caching code.

File: src/vm/util.cpp [hunk 5]
```diff
@@ -1861,8 +1861,6 @@ size_t GetLargestOnDieCacheSize(BOOL bTrueSize)
...
-
     static size_t maxSize;
     static size_t maxTrueSize;
```
Comment (sdmaclea) [inline src/vm/util.cpp]: Should these be using VolatileStore?  They were not so I left it alone

**Comment by sdmaclea** [comment]
@janvorli @Maoni0 This is an draft of the alternative change requested by @Maoni0 in #12771.  PTAL.  Untested, testing in progress.

**Comment by janvorli** [comment]
@sdmaclea I think that @Maoni0's idea was to get the cache size using processor specific asm code, something like we do for intel processors using the CPUID. But as far as I remember, you cannot do that on ARM / ARM64 in user mode, right? I guess she may not be aware of that.
As for reading the cache size from the pseudo file system, I actually think it is supported on Intel processors only and there we already have the _SC_LEVEL_XXXX stuff, so it seems it doesn't make sense to add another mean to get the same values.

**Comment by sdmaclea** [comment]
@janvorli Yes ccsidr_el1 & midr_el1 are not available from user space.

ARM has pushed 
```
commit 246246cbde5e840012f853e27630ebb59f409486
Author: Sudeep Holla <sudeep.holla@arm.com>
Date:   Tue Sep 30 14:48:25 2014 +0100

    drivers: base: support cpu cache information interface to userspace via sysfs
```

It is in the Xenial arm64 kernel source.  The device trees need to be updated to get this to work.  It doesn't seem to be working yet on many/most arm64 systems, but it is coming.  It will probably arrive before `_SC_LEVEL2_CACHE_SIZE` et. al. support, which may or may not make it to the Ubuntu 18.04 LTS initial release.  While I was here I wanted to pick up the info.  As kernel changes will be picked up quicker.

In any case I think this is currently the best I can do for implementing this interface.  Provided as an alternative

**Comment by janvorli** [comment]
@sdmaclea ah, I didn't know the support was coming. Then it makes sense to add this.

**Comment by sdmaclea** [comment]
I moved the sysfs read inside _ARM64_ because qemu was segfaulting on arm32.

I ran a full regression on Ubuntu arm64 and have stepped through the code in gdb it looks good.

@janvorli @Maoni0 let me know if you want changes

File: src/pal/src/misc/sysinfo.cpp (hunk 2 above)
Comment (sdmaclea) [inline src/pal/src/misc/sysinfo.cpp]: @janvorli Done

File: src/pal/src/misc/sysinfo.cpp [hunk 6]
```diff
@@ -403,5 +450,47 @@ PAL_GetLogicalProcessorCacheSizeFromOS()
     cacheSize = max(cacheSize, sysconf(_SC_LEVEL4_CACHE_SIZE));
 #endif
 
+    if(cacheSize == 0)
```
Comment (sdmaclea) [inline src/pal/src/misc/sysinfo.cpp]: @janvorli done

**Comment by sdmaclea** [comment]
@janvorli I think all your requested changes have been resolved.

File: src/pal/src/misc/sysinfo.cpp [hunk 7]
```diff
@@ -384,6 +384,53 @@ PAL_HasGetCurrentProcessorNumber()
     return HAVE_SCHED_GETCPU;
 }
 
+bool
+PALAPI
+ReadMemoryValueFromFile(const char* filename, size_t* val)
```
Comment (janvorli) [inline src/pal/src/misc/sysinfo.cpp]: A last nit - please remove the PALAPI, it is not an API exposed by PAL.

Comment (sdmaclea) [inline src/pal/src/misc/sysinfo.cpp]: @janvorli Done

**Comment by sdmaclea** [comment]
@dotnet-bot test Windows_NT x64 Formatting

(timeout)

**Comment by sdmaclea** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by sdmaclea** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by Maoni0** [comment]
Yes, I am fine with reading from the device file for now but I am counting on @janvorli to make sure the way it's reading from the file is reasonable as I'm only very vaguely familiar with it 😃

File: src/vm/util.cpp (hunk 5 above)
Comment (janvorli) [inline src/vm/util.cpp]: There is no need for volatile, as static local variables are guaranteed to be initialized in a thread safe manner. See http://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

**Pull request merged**